src/stream/src/executor/source/fetch_executor.rs

```diff
@@ -0,0 +1,467 @@
+        let fs_split = state_cache // Peek into state cache first.
+            .iter()
+            .find(|(_, split)| {
+                let fs_split = split.as_fs().unwrap();
+                fs_split.offset < fs_split.size
+            })
+            .map(|(_, split)| split.as_fs().unwrap().to_owned())
+            .or(loop {
+                // Otherwise find the next assignment in the state table.
+                if let Some(item) = state_table_iter.next().await {
+                    let row = item?;
+                    let split_id = match row.datum_at(0) {
+                        Some(ScalarRefImpl::Utf8(split_id)) => split_id,
+                        _ => unreachable!(),
+                    };
+
+                    // The state cache holds the latest status of the split.
+                    // Entering this branch means in the state cache offset >= size.
+                    // Hence we skip this split in the state table.
+                    if state_cache.contains_key(split_id) {
+                        continue;
+                    }
+
+                    let fs_split = match row.datum_at(1) {
+                        Some(ScalarRefImpl::Jsonb(jsonb_ref)) => {
+                            SplitImpl::restore_from_json(jsonb_ref.to_owned_scalar())?
+                                .as_fs()
+                                .unwrap()
+                                .to_owned()
+                        }
+                        _ => unreachable!(),
+                    };
+
+                    // Cache the assignment retrieved from state table.
+                    state_cache.insert(split_id.into(), fs_split.clone().into());
+                    break Some(fs_split);
+                }
+
+                break None;
+            });
+
+        if let Some(fs_split) = fs_split {
+            stream.replace_data_stream(
+                Self::build_stream_source_reader(column_ids, source_ctx, source_desc, fs_split)
+                    .await?,
+            );
+            *is_datastream_empty = false;
+        } else {
+            stream.replace_data_stream(stream::pending().boxed());
+            *is_datastream_empty = true;
+        };
+
+        Ok(())
+    }
+
+    async fn take_snapshot_and_flush(
+        state_store_handler: &mut SourceStateTableHandler<S>,
+        state_cache: &mut HashMap<SplitId, SplitImpl>,
+        epoch: EpochPair,
+    ) -> StreamExecutorResult<()> {
+        let mut to_flush = Vec::new();
+        let mut to_delete = Vec::new();
+        state_cache.iter().for_each(|(_, split)| {
+            let fs_split = split.as_fs().unwrap();
+            if fs_split.offset >= fs_split.size {
+                // If read out, try delete in the state table
+                to_delete.push(split.to_owned());
+            } else {
+                // Otherwise, flush to state table
+                to_flush.push(split.to_owned());
+            }
+        });
+        state_cache.clear();
```

Comment:
is it really necessary to clear all cache? how about just deleting the finished part